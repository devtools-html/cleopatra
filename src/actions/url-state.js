/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
