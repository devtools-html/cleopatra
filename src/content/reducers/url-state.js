// @flow
import { combineReducers } from 'redux';
import { defaultThreadOrder } from '../profile-data';
import { createSelector } from 'reselect';
import { urlFromState } from '../url-handling';
import * as RangeFilters from '../range-filters';

import type { ThreadIndex } from '../../common/types/profile';
import type { StartEndRange } from '../../common/types/units';
import type { Action, CallTreeFiltersPerThread, CallTreeFilter, DataSource } from '../actions/types';
import type { State, URLState, Reducer } from './types';

function dataSource(state: DataSource = 'none', action: Action) {
  switch (action.type) {
    case 'WAITING_FOR_PROFILE_FROM_FILE':
      return 'from-file';
    case 'PROFILE_PUBLISHED':
      return 'public';
    default:
      return state;
  }
}

function hash(state: string = '', action: Action) {
  switch (action.type) {
    case 'PROFILE_PUBLISHED':
      return action.hash;
    default:
      return state;
  }
}

function selectedTab(state: string = 'calltree', action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_TAB':
      return action.selectedTab;
    default:
      return state;
  }
}

function rangeFilters(state: StartEndRange[] = [], action: Action) {
  switch (action.type) {
    case 'ADD_RANGE_FILTER': {
      const { start, end } = action;
      return [...state, { start, end }];
    }
    case 'POP_RANGE_FILTERS':
      return state.slice(0, action.firstRemovedFilterIndex);
    default:
      return state;
  }
}

function selectedThread(state: ThreadIndex = 0, action: Action) {
  switch (action.type) {
    case 'CHANGE_SELECTED_THREAD':
      return action.selectedThread;
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_FILE': {
      // When loading in a brand new profile, select either the GeckoMain [tab] thread,
      // or the first thread in the thread order. For profiles from the Web, the
      // selectedThread has already been initialized from the URL and does not require
      // looking at the profile.
      const contentThreadId = action.profile.threads.findIndex(thread => {
        return thread.name === 'GeckoMain' && thread.processType === 'tab';
      });
      return contentThreadId !== -1 ? contentThreadId : defaultThreadOrder(action.profile.threads)[0];
    }
    default:
      return state;
  }
}

function callTreeSearchString(state: string = '', action: Action) {
  switch (action.type) {
    case 'CHANGE_CALL_TREE_SEARCH_STRING':
      return action.searchString;
    default:
      return state;
  }
}

function callTreeFilters(state: CallTreeFiltersPerThread = {}, action: Action) {
  switch (action.type) {
    case 'ADD_CALL_TREE_FILTER': {
      const { threadIndex, filter } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: [...oldFilters, filter],
      });
    }
    case 'POP_CALL_TREE_FILTERS': {
      const { threadIndex, firstRemovedFilterIndex } = action;
      const oldFilters = state[threadIndex] || [];
      return Object.assign({}, state, {
        [threadIndex]: oldFilters.slice(0, firstRemovedFilterIndex),
      });
    }
    default:
      return state;
  }
}

function jsOnly(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_JS_ONLY':
      return action.jsOnly;
    default:
      return state;
  }
}

function invertCallstack(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
      return action.invertCallstack;
    default:
      return state;
  }
}

function hidePlatformDetails(state: boolean = false, action: Action) {
  switch (action.type) {
    case 'CHANGE_HIDE_PLATFORM_DETAILS':
      return action.hidePlatformDetails;
    default:
      return state;
  }
}

const urlStateReducer: Reducer<URLState> = (regularUrlStateReducer => (state: URLState, action: Action): URLState => {
  switch (action.type) {
    case '@@urlenhancer/updateURLState':
      return action.urlState;
    default:
      return regularUrlStateReducer(state, action);
  }
})(combineReducers({
  dataSource, hash, selectedTab, rangeFilters, selectedThread,
  callTreeSearchString, callTreeFilters, jsOnly, invertCallstack,
  hidePlatformDetails,
}));
export default urlStateReducer;

const getURLState = (state: State): URLState => state.urlState;

export const getDataSource = (state: State) => getURLState(state).dataSource;
export const getHash = (state: State) => getURLState(state).hash;
export const getRangeFilters = (state: State) => getURLState(state).rangeFilters;
export const getJSOnly = (state: State) => getURLState(state).jsOnly;
export const getHidePlatformDetails = (state: State) => getURLState(state).hidePlatformDetails;
export const getInvertCallstack = (state: State) => getURLState(state).invertCallstack;
export const getSearchString = (state: State) => getURLState(state).callTreeSearchString;
export const getSelectedTab = (state: State) => getURLState(state).selectedTab;
export const getSelectedThreadIndex = (state: State) => getURLState(state).selectedThread;
export const getCallTreeFilters = (state: State, threadIndex: ThreadIndex): CallTreeFilter[] => {
  return getURLState(state).callTreeFilters[threadIndex] || [];
};

export const getURLPredictor = createSelector(
  getURLState,
  (oldURLState: URLState) => actionOrActionList => {
    const actionList = ('type' in actionOrActionList) ? [actionOrActionList] : actionOrActionList;
    const newURLState = actionList.reduce(urlStateReducer, oldURLState);
    return urlFromState(newURLState);
  }
);

export const getRangeFilterLabels = createSelector(
  getRangeFilters,
  RangeFilters.getRangeFilterLabels
);
