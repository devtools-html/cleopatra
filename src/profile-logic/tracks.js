/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import type { ScreenshotPayload } from '../types/markers';
import type { Profile, Thread, ThreadIndex, Pid } from '../types/profile';
import type {
  GlobalTrack,
  LocalTrack,
  TrackIndex,
} from '../types/profile-derived';
import { defaultThreadOrder, getFriendlyThreadName } from './profile-data';
import { ensureExists, assertExhaustiveCheck } from '../utils/flow';

/**
 * This file collects all the logic that goes into validating URL-encoded view options.
 * It also selects the default view options for things like track hiding, ordering,
 * and selection.
 */
export function initializeLocalTrackOrderByPid(
  urlTrackOrderByPid: Map<Pid, TrackIndex[]> | null,
  localTracksByPid: Map<Pid, LocalTrack[]>,
  legacyThreadOrder: ThreadIndex[] | null
): Map<Pid, TrackIndex[]> {
  const trackOrderByPid = new Map();

  if (legacyThreadOrder === null) {
    // Go through each set of tracks, determine the sort order.
    for (const [pid, tracks] of localTracksByPid) {
      // Create the default trackOrder.
      let trackOrder = tracks.map((_, index) => index);

      if (urlTrackOrderByPid !== null) {
        // Sanitize the track information provided by the URL, and ensure it is valid.
        const urlTrackOrder = urlTrackOrderByPid.get(pid);
        if (
          urlTrackOrder !== undefined &&
          _indexesAreValid(tracks.length, urlTrackOrder)
        ) {
          trackOrder = urlTrackOrder;
        }
      }

      trackOrderByPid.set(pid, trackOrder);
    }
  } else {
    // Convert the legacy thread order into the current track order.
    for (const [pid, tracks] of localTracksByPid) {
      const trackOrder = [];
      // Go through the legacy thread order and pair it with the correct track.
      for (const threadIndex of legacyThreadOrder) {
        const trackIndex = tracks.findIndex(
          track => track.type === 'thread' && track.threadIndex === threadIndex
        );
        if (trackIndex !== -1) {
          trackOrder.push(trackIndex);
        }
      }
      // Complete the list of track indexes by adding them to the end.
      for (let trackIndex = 0; trackIndex < tracks.length; trackIndex++) {
        if (!trackOrder.includes(trackIndex)) {
          trackOrder.push(trackIndex);
        }
      }
      trackOrderByPid.set(pid, trackOrder);
    }
  }

  return trackOrderByPid;
}

export function initializeHiddenLocalTracksByPid(
  urlHiddenTracksByPid: Map<Pid, Set<TrackIndex>> | null,
  localTracksByPid: Map<Pid, LocalTrack[]>,
  profile: Profile,
  legacyHiddenThreads: ThreadIndex[] | null
): Map<Pid, Set<TrackIndex>> {
  const hiddenTracksByPid = new Map();

  // Go through each set of tracks, determine the sort order.
  for (const [pid, tracks] of localTracksByPid) {
    const hiddenTracks = new Set();

    if (legacyHiddenThreads !== null) {
      for (const threadIndex of legacyHiddenThreads) {
        const trackIndex = tracks.findIndex(
          track => track.type === 'thread' && track.threadIndex === threadIndex
        );
        if (trackIndex !== -1) {
          hiddenTracks.add(trackIndex);
        }
      }
    } else if (urlHiddenTracksByPid === null) {
      for (let trackIndex = 0; trackIndex < tracks.length; trackIndex++) {
        const track = tracks[trackIndex];
        if (
          track.type === 'thread' &&
          _isThreadIdle(profile, profile.threads[track.threadIndex])
        ) {
          hiddenTracks.add(trackIndex);
        }
      }
    } else {
      const urlHiddenTracks = urlHiddenTracksByPid.get(pid);
      const trackIndexes = tracks.map((_, index) => index);

      if (urlHiddenTracks !== undefined) {
        for (const index of urlHiddenTracks) {
          if (trackIndexes.includes(index)) {
            hiddenTracks.add(index);
          }
        }
      }
    }

    hiddenTracksByPid.set(pid, hiddenTracks);
  }

  return hiddenTracksByPid;
}

export function computeLocalTracksByPid(
  profile: Profile
): Map<Pid, LocalTrack[]> {
  const localTracksByPid = new Map();

  for (
    let threadIndex = 0;
    threadIndex < profile.threads.length;
    threadIndex++
  ) {
    const thread = profile.threads[threadIndex];
    const { pid } = thread;
    // Get or create the tracks and trackOrder.
    let tracks = localTracksByPid.get(pid);
    if (tracks === undefined) {
      tracks = [];
      localTracksByPid.set(pid, tracks);
    }

    if (!_isMainThread(thread)) {
      // This thread has not been added as a GlobalTrack, so add it as a local track.
      tracks.push({ type: 'thread', threadIndex });
    }

    if (thread.markers.data.some(datum => datum && datum.type === 'Network')) {
      // This thread has network markers.
      tracks.push({ type: 'network', threadIndex });
    }
  }

  return localTracksByPid;
}

export function computeGlobalTracks(profile: Profile): GlobalTrack[] {
  // Defining this ProcessTrack type here helps flow understand the intent of
  // the internals of this function, otherwise each GlobalTrack usage would need
  // to check that it's a process type.
  type ProcessTrack = {
    type: 'process',
    pid: Pid,
    mainThreadIndex: number | null,
  };
  const globalTracksByPid: Map<Pid, ProcessTrack> = new Map();
  const globalTracks: GlobalTrack[] = [];

  // Create the global tracks.
  for (
    let threadIndex = 0;
    threadIndex < profile.threads.length;
    threadIndex++
  ) {
    const thread = profile.threads[threadIndex];
    const { pid, markers, stringTable } = thread;
    if (_isMainThread(thread)) {
      // This is a main thread, a global track needs to be created or updated with
      // the main thread info.
      let globalTrack = globalTracksByPid.get(pid);
      if (globalTrack === undefined) {
        // Create the track.
        globalTrack = {
          type: 'process',
          pid,
          mainThreadIndex: threadIndex,
        };
        globalTracks.push(globalTrack);
        globalTracksByPid.set(pid, globalTrack);
      } else {
        // The main thread index was found, add it.
        globalTrack.mainThreadIndex = threadIndex;
      }
    } else {
      // This is a non-main thread.
      if (!globalTracksByPid.has(pid)) {
        // This is a thread without a known main thread. Create a global process
        // track for it, but don't add a main thread for it.
        const globalTrack = {
          type: 'process',
          pid: pid,
          mainThreadIndex: null,
        };
        globalTracks.push(globalTrack);
        globalTracksByPid.set(pid, globalTrack);
      }
    }

    // Check for screenshots.
    const ids: Set<string> = new Set();
    if (stringTable.hasString('CompositorScreenshot')) {
      const screenshotNameIndex = stringTable.indexForString(
        'CompositorScreenshot'
      );
      for (let markerIndex = 0; markerIndex < markers.length; markerIndex++) {
        if (markers.name[markerIndex] === screenshotNameIndex) {
          // Coerce the payload to a screenshot one. Don't do a runtime check that
          // this is correct.
          const data: ScreenshotPayload = (markers.data[markerIndex]: any);
          ids.add(data.windowID);
        }
      }
      for (const id of ids) {
        globalTracks.unshift({ type: 'screenshots', id, threadIndex });
      }
    }
  }
  return globalTracks;
}

export function initializeGlobalTrackOrder(
  globalTracks: GlobalTrack[],
  urlGlobalTrackOrder: TrackIndex[] | null,
  legacyThreadOrder: ThreadIndex[] | null
): TrackIndex[] {
  if (legacyThreadOrder !== null) {
    // Upgrade an older URL value based on the thread index to the track index based
    // ordering. Don't trust that the thread indexes are actually valid.
    const trackOrder = [];

    // Convert the thread index to a track index, if it's valid.
    for (const threadIndex of legacyThreadOrder) {
      const trackIndex = globalTracks.findIndex(
        globalTrack =>
          globalTrack.type === 'process' &&
          globalTrack.mainThreadIndex === threadIndex
      );
      if (trackIndex !== -1) {
        trackOrder.push(trackIndex);
      }
    }

    // Add the remaining track indexes.
    for (let trackIndex = 0; trackIndex < globalTracks.length; trackIndex++) {
      if (!trackOrder.includes(trackIndex)) {
        trackOrder.push(trackIndex);
      }
    }
    return trackOrder;
  }

  return urlGlobalTrackOrder !== null &&
    _indexesAreValid(globalTracks.length, urlGlobalTrackOrder)
    ? urlGlobalTrackOrder
    : globalTracks.map((_, index) => index);
}

export function initializeSelectedThreadIndex(
  selectedThreadIndex: ThreadIndex | null,
  visibleThreadIndexes: ThreadIndex[],
  profile: Profile
): ThreadIndex | null {
  if (
    selectedThreadIndex !== null &&
    visibleThreadIndexes.includes(selectedThreadIndex)
  ) {
    // This is a valid thread index to select.
    return selectedThreadIndex;
  }
  // Select either the GeckoMain [tab] thread, or the first thread in the thread
  // order.
  const threadIndex = profile.threads.indexOf(
    _findDefaultThread(
      visibleThreadIndexes.map(threadIndex => profile.threads[threadIndex])
    )
  );
  // The threadIndex being null should only happen in tests where profiles
  // have no threads.
  return threadIndex === -1 ? null : threadIndex;
}

export function initializeHiddenGlobalTracks(
  globalTracks: GlobalTrack[],
  profile: Profile,
  validTrackIndexes: TrackIndex[],
  urlHiddenGlobalTracks: Set<TrackIndex> | null,
  legacyHiddenThreads: ThreadIndex[] | null
): Set<TrackIndex> {
  const hiddenGlobalTracks = new Set();

  if (legacyHiddenThreads !== null) {
    for (const threadIndex of legacyHiddenThreads) {
      const trackIndex = globalTracks.findIndex(
        track =>
          track.type === 'process' && track.mainThreadIndex === threadIndex
      );
      if (trackIndex !== -1) {
        hiddenGlobalTracks.add(trackIndex);
      }
    }
  }

  if (urlHiddenGlobalTracks === null) {
    // No hidden global tracks exist, generate the default Set.
    for (let trackIndex = 0; trackIndex < globalTracks.length; trackIndex++) {
      const track = globalTracks[trackIndex];
      if (track.type === 'process' && track.mainThreadIndex !== null) {
        const { mainThreadIndex } = track;
        const thread = profile.threads[mainThreadIndex];
        if (_isThreadIdle(profile, thread)) {
          hiddenGlobalTracks.add(trackIndex);
        }
      }
    }
    return hiddenGlobalTracks;
  }

  // Validate the global tracks to hide.
  for (const trackIndex of urlHiddenGlobalTracks) {
    if (validTrackIndexes.includes(trackIndex)) {
      hiddenGlobalTracks.add(trackIndex);
    }
  }
  return hiddenGlobalTracks;
}

export function getVisibleThreads(
  globalTracks: GlobalTrack[],
  hiddenGlobalTracks: Set<TrackIndex>,
  localTracksByPid: Map<Pid, LocalTrack[]>,
  hiddenTracksByPid: Map<Pid, Set<TrackIndex>>
): ThreadIndex[] {
  const visibleThreads = [];
  for (const globalTrack of globalTracks) {
    if (globalTrack.type === 'process') {
      const { mainThreadIndex, pid } = globalTrack;
      if (
        mainThreadIndex !== null &&
        !hiddenGlobalTracks.has(mainThreadIndex)
      ) {
        visibleThreads.push(mainThreadIndex);
      }
      const tracks = ensureExists(
        localTracksByPid.get(pid),
        'A local track was expected to exist for the given pid.'
      );
      const hiddenTracks = ensureExists(
        hiddenTracksByPid.get(pid),
        'Hidden tracks were expected to exists for the given pid.'
      );
      for (const track of tracks) {
        if (track.type === 'thread') {
          const { threadIndex } = track;
          if (!hiddenTracks.has(threadIndex)) {
            visibleThreads.push(threadIndex);
          }
        }
      }
    }
  }
  return visibleThreads;
}

export function getGlobalTrackName(
  globalTrack: GlobalTrack,
  threads: Thread[]
): string {
  switch (globalTrack.type) {
    case 'process': {
      // Look up the thread information for the process if it exists.
      if (globalTrack.mainThreadIndex === null) {
        // No main thread was found for process track, so it is empty. This can
        // happen for instance when recording "DOM Worker" but not "GeckoMain". The
        // "DOM Worker" thread will be captured, but not the main thread, thus leaving
        // a process track with no main thread.
        return typeof globalTrack.pid === 'string'
          ? // The pid is a unique string label, use that.
            globalTrack.pid
          : // The pid is a number, make a label for it.
            `Process ${globalTrack.pid}`;
      }
      return getFriendlyThreadName(
        threads,
        threads[globalTrack.mainThreadIndex]
      );
    }
    case 'screenshots':
      return 'Screenshots';
    default:
      throw assertExhaustiveCheck(globalTrack, 'Unhandled GlobalTrack type.');
  }
}

export function getLocalTrackName(
  localTrack: LocalTrack,
  threads: Thread[]
): string {
  switch (localTrack.type) {
    case 'thread':
      return getFriendlyThreadName(threads, threads[localTrack.threadIndex]);
    case 'network':
      return 'Network';
    case 'memory':
      return 'Memory';
    default:
      throw assertExhaustiveCheck(localTrack, 'Unhandled LocalTrack type.');
  }
}

// Any thread with less than 1% non-idle time will be hidden.
const PERCENTAGE_ACTIVE_SAMPLES = 0.01;

/**
 * Determine if a thread is idle, so that it can be hidden. It is really annoying for an
 * end user to load a profile full of empty and idle threads. This function goes through
 * all of the samples in the thread, and sees if some large percentage of them are idle.
 */
function _isThreadIdle(profile: Profile, thread: Thread): boolean {
  if (
    // Don't hide the compositor.
    thread.name === 'Compositor' ||
    // Don't hide the main thread.
    (thread.name === 'GeckoMain' && thread.processType === 'default') ||
    // Don't hide the GPU thread on Windows.
    (thread.name === 'GeckoMain' && thread.processType === 'gpu')
  ) {
    return false;
  }

  let maxActiveStackCount = PERCENTAGE_ACTIVE_SAMPLES * thread.samples.length;
  let activeStackCount = 0;
  let filteredStackCount = 0;

  for (
    let sampleIndex = 0;
    sampleIndex < thread.samples.length;
    sampleIndex++
  ) {
    const stackIndex = thread.samples.stack[sampleIndex];
    if (stackIndex === null) {
      // This stack was filtered out. Most likely this will never actually happen
      // on a new profile, but keep this check here since the stacks are possibly
      // null in the Flow type definitions.
      filteredStackCount++;
      // Adjust the maximum necessary active stacks to find based on null stacks.
      maxActiveStackCount =
        PERCENTAGE_ACTIVE_SAMPLES *
        (thread.samples.length - filteredStackCount);
    } else {
      const categoryIndex = thread.stackTable.category[stackIndex];
      const category = profile.meta.categories[categoryIndex];
      if (category.name !== 'Idle') {
        activeStackCount++;
        if (activeStackCount > maxActiveStackCount) {
          return false;
        }
      }
    }
  }

  // Do one final check to see if we have enough active samples.
  return activeStackCount <= maxActiveStackCount;
}

function _findDefaultThread(threads: Thread[]): Thread | null {
  if (threads.length === 0) {
    // Tests may have no threads.
    return null;
  }
  const contentThreadId = threads.findIndex(
    thread => thread.name === 'GeckoMain' && thread.processType === 'tab'
  );
  const defaultThreadIndex =
    contentThreadId !== -1 ? contentThreadId : defaultThreadOrder(threads)[0];

  return threads[defaultThreadIndex];
}

function _isMainThread(thread: Thread): boolean {
  return (
    thread.name === 'GeckoMain' ||
    // If the pid is a string, then it's not one that came from the system.
    // These threads should all be treated as main threads.
    typeof thread.pid === 'string'
  );
}

function _indexesAreValid(listLength: number, indexes: number[]) {
  return (
    // The item length is valid.
    indexes.length === listLength &&
    // The indexes are valid and include every single value.
    indexes
      .slice()
      .sort()
      .every((value, arrayIndex) => value === arrayIndex)
  );
}
