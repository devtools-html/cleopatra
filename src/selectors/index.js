/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import { selectedThreadSelectors as selectedThread } from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';

// Normally we don't need that type because flow can infer types, but on that
// object, it gives weird flow errors that say missing type annotation for X.
// Only adding a type like that seems to solve it.
type Selectors = {
  app: typeof app,
  profile: typeof profile,
  urlState: typeof urlState,
  icons: typeof icons,
  publish: typeof publish,
  zippedProfiles: typeof zippedProfiles,
  selectedThread: typeof selectedThread,
};

export default ({
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
}: Selectors);
