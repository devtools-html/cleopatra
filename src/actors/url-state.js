/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
