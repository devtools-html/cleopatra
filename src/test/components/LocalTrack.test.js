/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type { TrackReference } from '../../types/actions';
import type { Store } from '../../types/store';
import type { LocalTrack } from '../../types/profile-derived';

import * as React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import {
  changeSelectedThread,
  hideLocalTrack,
} from '../../actions/profile-view';
import TimelineLocalTrack from '../../components/timeline/LocalTrack';
import {
  getRightClickedTrack,
  getLocalTrackFromReference,
  getProfile,
} from '../../selectors/profile';
import { getSelectedThreadIndex } from '../../selectors/url-state';
import mockCanvasContext from '../fixtures/mocks/canvas-context';
import { getNetworkTrackProfile } from '../fixtures/profiles/make-profile';
import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';
import { storeWithProfile } from '../fixtures/stores';

// In getProfileWithNiceTracks, the two pids are 111 and 222 for the
// "GeckoMain process" and "GeckoMain tab" respectively. Use 222 since it has
// local tracks.
const PID = 222;
const LEFT_CLICK = 0;
const RIGHT_CLICK = 2;

describe('timeline/LocalTrack', function() {
  describe('with a thread track', function() {
    it('matches the snapshot of a local track', () => {
      const { view } = setupThreadTrack();
      expect(view).toMatchSnapshot();
    });

    it('has the correct selectors into useful parts of the component', function() {
      const { getLocalTrackLabel, getLocalTrackRow } = setupThreadTrack();
      expect(getLocalTrackLabel().text()).toBe('DOM Worker');
      expect(getLocalTrackRow().exists()).toBe(true);
    });

    it('starts out not being selected', function() {
      const {
        getState,
        threadIndex,
        trackReference,
        getLocalTrackRow,
      } = setupThreadTrack();
      expect(getRightClickedTrack(getState())).not.toEqual(trackReference);
      expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
      expect(getLocalTrackRow().hasClass('selected')).toBe(false);
    });

    it('can select a thread by clicking the label', () => {
      const {
        getState,
        getLocalTrackLabel,
        threadIndex,
        getLocalTrackRow,
      } = setupThreadTrack();
      expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
      getLocalTrackLabel().simulate('mousedown', { button: LEFT_CLICK });
      expect(getSelectedThreadIndex(getState())).toBe(threadIndex);
      expect(getLocalTrackRow().hasClass('selected')).toBe(true);
    });

    it('can right click a thread', () => {
      const {
        getState,
        getLocalTrackLabel,
        threadIndex,
        trackReference,
      } = setupThreadTrack();

      getLocalTrackLabel().simulate('mousedown', { button: RIGHT_CLICK });
      expect(getRightClickedTrack(getState())).toEqual(trackReference);
      expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
    });

    it('can select a thread by clicking the row', () => {
      const { getState, getLocalTrackRow, threadIndex } = setupThreadTrack();
      expect(getSelectedThreadIndex(getState())).not.toBe(threadIndex);
      getLocalTrackRow().simulate('click');
      expect(getSelectedThreadIndex(getState())).toBe(threadIndex);
    });

    it('will render a stub div if the track is hidden', () => {
      const { view, pid, trackReference, dispatch } = setupThreadTrack();
      dispatch(hideLocalTrack(pid, trackReference.trackIndex));
      view.update();
      expect(view.find('.timelineTrackHidden').exists()).toBe(true);
      expect(view.find('.timelineTrack').exists()).toBe(false);
    });
  });

  describe('with a network track', function() {
    it('has correctly renders the network label', function() {
      const { getLocalTrackLabel } = setupWithNetworkProfile();
      expect(getLocalTrackLabel().text()).toBe('Network');
    });

    it('matches the snapshot of the network track', () => {
      const { view } = setupWithNetworkProfile();
      expect(view).toMatchSnapshot();
    });
  });
});

function setup(
  store: Store,
  trackReference: TrackReference,
  localTrack: LocalTrack
) {
  const { getState, dispatch } = store;
  const { threadIndex } = localTrack;
  // The assertions are simpler if this thread is not already selected.
  dispatch(changeSelectedThread(threadIndex + 1));

  // Some child components render to canvas.
  jest
    .spyOn(HTMLCanvasElement.prototype, 'getContext')
    .mockImplementation(() => mockCanvasContext());

  const view = mount(
    <Provider store={store}>
      <TimelineLocalTrack
        pid={PID}
        localTrack={localTrack}
        trackIndex={trackReference.trackIndex}
      />
    </Provider>
  );

  const getLocalTrackLabel = () => view.find('.timelineTrackLabel').first();
  const getLocalTrackRow = () => view.find('.timelineTrackLocalRow').first();

  return {
    dispatch,
    getState,
    profile: getProfile(getState()),
    store,
    view,
    trackReference,
    threadIndex,
    pid: PID,
    getLocalTrackLabel,
    getLocalTrackRow,
  };
}

/**
 *  getProfileWithNiceTracks() looks like: [
 *    'show [thread GeckoMain process]',
 *    'show [thread GeckoMain tab]',
 *    '  - show [thread DOM Worker]',         <- use this local track.
 *    '  - show [thread Style]',
 *  ]
 */
function setupThreadTrack() {
  // Select the first local track index, which should be the one noted in the
  // the comment above.
  const trackIndex = 0;
  const profile = getProfileWithNiceTracks();
  const store = storeWithProfile(profile);
  const trackReference = { type: 'local', pid: PID, trackIndex };
  const localTrack = getLocalTrackFromReference(
    store.getState(),
    trackReference
  );
  if (localTrack.type !== 'thread') {
    throw new Error('Expected a thread track.');
  }
  return setup(store, trackReference, localTrack);
}

function setupWithNetworkProfile() {
  // Select the 2nd track, which will be the network track, while the first is the
  // main thread.
  const trackIndex = 1;
  const profile = getNetworkTrackProfile();
  profile.threads[0].pid = PID;

  const store = storeWithProfile(profile);
  const trackReference = { type: 'local', pid: PID, trackIndex };
  const localTrack = getLocalTrackFromReference(
    store.getState(),
    trackReference
  );
  if (localTrack.type !== 'network') {
    throw new Error('Expected a network track.');
  }
  return setup(store, trackReference, localTrack);
}
