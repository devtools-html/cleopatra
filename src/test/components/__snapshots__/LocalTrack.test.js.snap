// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/LocalTrack with a network track matches the snapshot of the network track 1`] = `
<li
  className="timelineTrack timelineTrackLocal"
>
  <div
    className="timelineTrackRow timelineTrackLocalRow"
    onClick={[Function]}
  >
    <div
      className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
      onContextMenu={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title={null}
    >
      <button
        className="timelineTrackNameButton"
        type="button"
      >
        Network
      </button>
    </div>
    <div
      className="timelineTrackTrack"
    >
      <div
        className="timelineTrackNetwork"
        style={
          Object {
            "height": 35,
          }
        }
      >
        <canvas
          className="timelineTrackNetworkCanvas"
        />
      </div>
    </div>
  </div>
</li>
`;

exports[`timeline/LocalTrack with a thread track matches the snapshot of a local track 1`] = `
<li
  className="timelineTrack timelineTrackLocal"
>
  <div
    className="timelineTrackRow timelineTrackLocalRow"
    onClick={[Function]}
  >
    <div
      className="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
      onContextMenu={[Function]}
      onMouseDown={[Function]}
      onMouseOut={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title="thread: \\"DOM Worker\\" (0)
process: \\"tab\\" (222)"
    >
      <button
        className="timelineTrackNameButton"
        type="button"
      >
        DOM Worker
      </button>
    </div>
    <div
      className="timelineTrackTrack"
    >
      <div
        className="timelineTrackThread"
      >
        <div
          className="timelineTrackThreadIntervalMarkerOverview"
        >
          <canvas
            className="timelineTracingMarkersCanvas"
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            onMouseOut={[Function]}
            onMouseUp={[Function]}
          />
        </div>
        <div
          className="threadActivityGraph"
        >
          <canvas
            className="threadActivityGraphCanvas threadActivityGraphCanvas"
            onMouseUp={[Function]}
          />
        </div>
        <div
          className="timelineEmptyThreadIndicator"
        />
      </div>
    </div>
  </div>
</li>
`;
